### PR TITLE
Settings redesign

### PR DESCRIPTION
**Pull request by pserwylo**
The rationale for this is that I've been motivated to return to #28. I have a working version, but the next step requires adding an option in the config screen, and it was going to be a bit messy to do it with the current custom config screen. By using the `androidx.preference` library (the default way of doing settings in Android apps), it will be trivial to:
* Add an option to allow auto play songs.
* Let the user select from a list of songs they want to be included in the playlist.
* Potentially other cool things such as giving the user the option to record their own songs.

The new version is backwards compatible (will show the correct sound set from the previously chosen one in the old config screen). It also adds the notion of `strings.xml` values for each soundset so that we can name them using human readable strings, instead of relying on the folders in the filesystem.

Here is a comparison of the old and new:

### Old

old

### New

new 
new2

It is worth noting that while testing the "Quit app" functionality, it crashes in the same way on current `master` with this:

```
2022-07-03 21:54:15.100 29362-29362/com.nicobrailo.pianoli E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.nicobrailo.pianoli, PID: 29362
    java.lang.RuntimeException: Unable to resume activity {com.nicobrailo.pianoli/com.nicobrailo.pianoli.MainActivity}: java.lang.IllegalArgumentException: Invalid task, not in foreground
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4906)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4949)
        at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:52)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:190)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:105)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2458)
        at android.os.Handler.dispatchMessage(Handler.java:110)
        at android.os.Looper.loop(Looper.java:219)
        at android.app.ActivityThread.main(ActivityThread.java:8387)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:513)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1055)
     Caused by: java.lang.IllegalArgumentException: Invalid task, not in foreground
        at android.os.Parcel.createException(Parcel.java:2075)
        at android.os.Parcel.readException(Parcel.java:2039)
        at android.os.Parcel.readException(Parcel.java:1987)
        at android.app.IActivityTaskManager$Stub$Proxy.startLockTaskModeByToken(IActivityTaskManager.java:5462)
        at android.app.Activity.startLockTask(Activity.java:8505)
        at com.nicobrailo.pianoli.MainActivity.lock_app(MainActivity.java:58)
        at com.nicobrailo.pianoli.MainActivity.onResume(MainActivity.java:70)
        at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1472)
        at android.app.Activity.performResume(Activity.java:8258)
        at android.app.ActivityThread.performResumeActivity(ActivityThread.java:4896)
        at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:4949) 
        at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:52) 
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:190) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:105) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2458) 
        at android.os.Handler.dispatchMessage(Handler.java:110) 
        at android.os.Looper.loop(Looper.java:219) 
        at android.app.ActivityThread.main(ActivityThread.java:8387) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:513) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1055) 
     Caused by: android.os.RemoteException: Remote stack trace:
        at com.android.server.wm.ActivityTaskManagerService.startLockTaskModeLocked(ActivityTaskManagerService.java:3265)
        at com.android.server.wm.ActivityTaskManagerService.startLockTaskModeByToken(ActivityTaskManagerService.java:3210)
        at android.app.IActivityTaskManager$Stub.onTransact(IActivityTaskManager.java:2530)
        at android.os.Binder.execTransactInternal(Binder.java:1028)
        at android.os.Binder.execTransact(Binder.java:1001)
```

However, this doesn't really matter all that much, because it still has the desired behaviour of closing the app :smile:.

### CODE REVIEW

File: app/src/main/java/com/nicobrailo/pianoli/SettingsActivity.java [hunk 1]
```diff
@@ -20,6 +20,8 @@
 import java.util.List;
 import java.util.Map;
 
+import kotlin.jvm.internal.PropertyReference0Impl;
```
Comment (nicolasbrailo) [inline app/src/main/java/com/nicobrailo/pianoli/SettingsActivity.java:23]: Is this imported on purpose? I'm not familiar with kotlin, but an import called "Reference0Impl" seems suspicious

**Review by nicolasbrailo — APPROVED** [review]

File: app/src/main/java/com/nicobrailo/pianoli/SettingsActivity.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        ArrayList<String> filtList = new ArrayList<>();
+        for (final String s : lst) {
+            if (s.startsWith(SOUNDSET_DIR_PREFIX)) {
+                // User display should be the asset name without the prefix
+                filtList.add(s.substring(SOUNDSET_DIR_PREFIX.length()));
+            }
+        }
+
+        if (filtList.size() == 0) {
+            final String msg = "No sounds found, the keyboard won't work!";
```
Comment (nicolasbrailo) [inline app/src/main/java/com/nicobrailo/pianoli/SettingsActivity.java:72]: Can we add a comment explaining a naive user what they should do? I'm thinking something like "This maybe a bug. Please try reinstalling the app. If that doesn't work, please report a bug at github.com/nicolasbrailo/PianOli"

File: app/src/main/java/com/nicobrailo/pianoli/SettingsActivity.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        return true;
+    }
+
+    @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+        if (item.getItemId() == R.id.menu_item_quit_app) {
+            onQuit();
+            return true;
+        }
+
```
Comment (nicolasbrailo) [inline app/src/main/java/com/nicobrailo/pianoli/SettingsActivity.java:99]: Does this need a setResult() with a different value?

**Review by nicolasbrailo — APPROVED** [review]

**Comment by nicolasbrailo** [comment]
Thank you for this change. It looks a lot cleaner.

In the screenshot, it isn't entirely clear to me how you'd close the app. Is the X on the top right meant to close the app, or just the settings window? Do you think that'd be confusing to an unfamiliar user?

> It is worth noting that while testing the "Quit app" functionality, it crashes in the same way on current master with this:

Lol. I guess that's one way of quitting...

**Comment by pserwylo** [comment]
> In the screenshot, it isn't entirely clear to me how you'd close the app. Is the X on the top right meant to close the app, or just the settings window? Do you think that'd be confusing to an unfamiliar user?

Yes, quite possibly. I can't seem to find a way to get the text next to the icon in portrait mode. One option is a different icon:

power_settings_new_FILL0_wght400_GRAD0_opsz48

... but that may not be obvious either.

Now that I think about it, I think we can relatively simply have the best of both worlds: Landscape view (with proper quit button, and also maintaining the same orientation as the keyboard and hence preventing the need to flip the phone), with a proper settings implementation. I could do this by choping up the new `SettingsActivity.SettingsFragment` and placing it on the right side, and leaving the header from my screenshots on the left side (or the right to be the same as before). 

I had a little play, not 100% sure it is better though:

Screenshot_20220704_205814

Screenshot_20220704_205404

new

Thoughts?

**Pull request merged**